### ISSUE TITLE
New Solar Enevelope by Boris Ready for Beta Testing

### ISSUE THREAD

**Issue by chriswmackey** [mostaphaRoudsari/ladybug#161]
I have started this discussion as @boris-p  has just about worked out the last of the major known bugs in the VERY AWESOME, VERY POWERFUL, new custom solar envelope component.  This will be the place where people who are interested in testing the component should post comments and feedback that will be taken into account before a complete component is released (hopefully before the next stable release of Ladybug).

**Comment by chriswmackey** [comment]
I post @boris-p 's most recent email outlining all of his most recent updates below (I will respond to these in the next post):

Hey Chris and Mostapha,
 
I apologize for the long delay (and the long email). Had some busy times and also wanted to address the issues with care before going further. In short, the biggest issue was the multithreading problem. I've made some changes and many tests. Tried really hard and still couldn't get it to crash (I'll elaborate on the process and the tests later in the email). From my side, I think now would be a good time to let some more people test it so we could get better feedback and see how this performs on other machines and with other models. In parallel I'll prepare a tutorial, either a video or a written one.
 
I saw that you have updated the license to GPL. Just to make sure, could you let me know exactly what I should add to the beginning of the file?
 
Perhaps it would be a good idea to move this conversation to github? Might be easier to follow and keep a track of everything. If you think it's a good idea I could open a new issue and post there.
 
A few changes to the last file that Chris has sent. Let me know if that makes sense for you  -
·        Regarding the max height - because it's opposite for solar envelope and solar collection (minimum or maximum value) I think that the message should also reflect that.
·        perhaps the max and min height should not be the same value in absolute terms. To explain -  I can imagine a default value of about 100 meters above the basesrf making sense but for the minimum I think something like  -20 would me more than enough and more than that might be confusing and look like an error.
·        One of the input parameters determines whether the component would run the standard solar envelope or the solar collection envelope, I think that for the user it would be more intuitive to make the selection as a dropdown list instead of a boolean (unless we make the input names really long and descriptive ). I have mapped the dropdown list values to - Solar Envelop = True, Solar Collection = False. As far as the component is concerned,  it's the same boolean, just think it would be more intuitive for the user.
·        I've incorporated Chris's improvements (thanks a lot, I really learned a lot in the process) but decided to stick with classes with the main reason being that I think for me it would be easier to maintain and extend. In doing so I have made many tests and comparisons between the files to check for stability and performance. I have found no performance differences, and will elaborate regarding stability (referring to multithreading ) in the next part, please let me know what you think.
 
Now regarding multithreading . Something strange is happening there that I still don't fully understand but I think that now the solution is stable. If I understood correctly we are using a python implementation or an api to a .net Threading library - https://msdn.microsoft.com/en-us/library/system.threading.tasks(v=vs.110).aspx
 
The thing that was strange for me from the beginning was that in essence there shouldn't be a case of a race condition because there is no overlapping of writing data. Different threads get different points to analyze and there is no change to common variables. I still don't understand  how we have a crash relating to Read operations, this should not happen and I think it is some sort of a bug caused by the Threading library making a lot of things automatically and in the process locks values or thinks that different threads are also writing and not just reading common values. This assumption is reinforced by Chris's solution, when we're copying the data (like the list of suns or obstacle curves) the chance for a collision is much smaller .
I tried to see how can we be really safe and decided to split the list of points to be checked to X sublists, X being the number of threads that is now manually decided upon (maybe should be an input parameter as well? in any case there's an option to run without threads). To every sublist make a copy of all the data that it uses.
I have used deepcopy to make sure that when I copy complex Objects they would still not reference the same values (and I think that if done this way, having classes should not be an issue).
When all the threads finish I join the sublists back to one list of points and continue synchronously.
 
I have set up an automated test in grasshopper which iterates the execution of the component and changes the input values (I changed one of the surrounding buildings, just to make sure that we don't run the component with the same data and to simulate user interaction better). So one iteration is execution and production of the solar envelope component. To give a sense of scale,  in my original component it would crash after 1-6 times! regularly.
So , for the results -
 
* All the tests were performed with a gridsize set to 3.
 
With Chris's improvements -
first run - crash on 56th iteration
second run - crash on 11th iteration
third run - crash on 135th iteration
forth run - crash on 20th iteration
 
 
When copying all the data for every thread except for the list of suns  -
first run  - crash on 35th  iteration
second run - crash on 120th iteration
third run  - crash on 235th iteration
forth run  - crash on 17th  iteration
fifth run  - crash on 177th iteration
* although we are just reading from the suns array it still crashes, very strange.
 
When copying all the data for every thread -
first run - manually stopped after more than 200 iterations
second run - manually stopped after more than 200 iterations
third run - manually stopped after more than 400 iterations
forth run - manually stopped after more than 400 iterations
 
 
Attached is the file I used for testing using a timer. Also streaming the number of iterations to an external file so that I could keep it running  without being there and still have a documentation as to when it (hopefully never) crashes.
 
I feel much better about this version now,  let me know what you think and we could get it out there.

**Comment by chriswmackey** [comment]
@boris-p ,

That is great to hear that you worked through the issue of the crashing parallel process.  Your rigor in testing it is truly commendable.

To go through your points:

Yes, we decided it was time to make the commitment to GPL after there seemed to be some companies with strong incentives to take the code of LB and and copy it into their closed-source programs in order to charge people for it.  If you sync your Ladybug with the github, you will see that @mostaphaRoudsari  updated the licence at the top of your previous version in the WIP section (you can see the code for it here:   https://github.com/mostaphaRoudsari/ladybug/blob/master/src/Ladybug_SolarEnvelopeAdvanced.py#L1-L23)

I hope you don't mind that I beat you to the github post.  I just thought it would be great to involve everyone in the discussion soon, particularly with how we implement this parallel process.

I think that all of your changes are great and, if I did not clarify last time when I sent you feedback, all of the edits that I made to your component were merely suggestions (and very hacky ones at that).  I know that I freak out sometimes when people edit my code too much and I respect your desire to keep things in a format that you are better able to understand and support.  I was just trying everything that I could to get the parallel process not to crash and, as you point out, some of the things were helpful (like copying the lists before the process) and some were not (like getting rid of the classes that are important to your navigation through your code).

Your tests are great and I have also not been able to get the parallel process to crash with your new component.  Because of the automated testing in the GH file that you sent Mostapha and I, when I open it, the component re-runs with new geometry every 2 seconds.  This results in an unending loop of running and re-running that makes Rhino/GH freeze every time that it runs.  Here's a visual of my computer's performance when I open your file:
![soler envelope process keeps re-running](https://cloud.githubusercontent.com/assets/5567574/9027367/0e7ffaf4-3921-11e5-8252-cb8fd4460dad.PNG)

Below I have included the exact same GH file that you sent Mostapha and I but I have stopped this automated testing to make it easier for other users to run the GH file.  I would suggest that people interested in beta testing use this until @boris-p posts and updated versions:
https://www.dropbox.com/s/jv0eifiwgp7g7m4/solarEnvelopeAdvanced_29.7_parralel.gh?dl=0
https://www.dropbox.com/s/7lkrq9bgfn1nq69/base_3d_file_29.7.3dm?dl=0

@boris-p , If you could send us an updated pull request with your new component, that would be great so that beta-testers can find it in their toolbar.

I am only going to give one suggestion and that is that you either allow people to set the number of CPU's and the chunking of the data manually (similarly to how we do it on the RunDaylightSimulation component):
![daylightcomponent](https://cloud.githubusercontent.com/assets/5567574/9027435/dcf4e4de-3922-11e5-934a-912f58777479.PNG)


Alternatively, you can try to read the number of CPUs on the machine and put in a value less than that.  I am in favor of a numberOfCPUs input with a default set to 1 but I think this is an issue that we should discuss it here. In any case, it definitely has to change from the current implementation that sets the numberOfCPUs to 4 as there are still some people running old-school single core machines out there (there were a couple in my Ladybug workshop this past week).  Also, it would be a shame if those with powerful computers would not take advantage of more cores.  I post your current way of assigning the default CPUs below:
![numberofprocessors](https://cloud.githubusercontent.com/assets/5567574/9027428/bb3df01a-3922-11e5-8add-31cb66284265.PNG)


I think that's it for now.  Let the beta testing and discussion begin!
-Chris

**Comment by boris-p** [comment]
@chriswmackey , I agree with all your comments. Didn't notice that I saved the file with the automated testing turned on so that it runs straight when you open the file, I would have been annoyed opening such a file but in addition this testing was more for us to really try and confront the problem, I think that most users, even testers don't really need to run it.
Couldn't agree more regarding the CPUs. I've added the number of CPUs as a parameter for now because although I've made some tests with automatically reading the number of CPUs, for some reason on my machine with 4 physical and 8 logical CPUs, the simulation actually runs fastest at 4 and gets slower when increasing the number of CPUs beyond that, which is strange by itself. 
I've added a test file with my latest updates - 
https://www.dropbox.com/s/3lrr1qmzbt2xm0i/solarEnvelopeAdvanced_2.8_test_case.gh?dl=0
Maybe we could copy the file to a more public location?

Also sent a pull request with the latest version.
Let the games begin?

**Comment by agusHariyadi** [comment]
@boris-p , @chriswmackey  , when I try to compare it in my computer, the finalPointList have different result with envelopePts.
https://www.dropbox.com/s/ifbufn026agvvc3/solarEnvelopeAdvanced_2.8_test_case_compare.gh?dl=0
![capture](https://cloud.githubusercontent.com/assets/6168707/9031287/ce1022b6-39eb-11e5-8364-1b529243d475.PNG)

**Comment by boris-p** [comment]
@agusHariyadi , thanks a lot for the feedback. The difference occurs due to the way that we calculate the maximum height of the points when they don't interfere with other buildings no matter how high they would be.The previous solution measured 100 above absolute zero while the new one measure 100 units above the base curve defined. in the example you could see that the base crv's z value is -4 so the heights are at 96. Also there's a parameter to manually set the maxHeight_ and if you set it to 104 all the points should align.

**Comment by ayezioro** [comment]
This is great Boris!! So similar to Guedi's original. Good contribution and hope our students will enjoy this.
I tried it a bit, and i have some comments. I see that the running time is already an issue. I tested your last example and changing the north orientation the time changes from 5 seconds to 15, depending on "how are your neighbors". Choosing "properly" the analysis period must be a careful issue that may be worth to test.
Not a big one, but if possible the "_obstacleCrvs" can be connected as "_context" (to be more consistent with other components), but instead of being curves they be breps. The process of transforming them to curves can be intern to your component. I understand that so far it is working for planar curves, right?
You wrote that the resulting points are not the same. Are you sure? I think they are the same but not in the same order (which can be weird, but possible). I identified many points in different order on both lists.
In the value list you are choosing the envelope you need to change "SolarRights" to "SolarEnvelope" to be consistent to the input.
I would also add the explanation you have at the top of your python file to the hint of the component, so users can get the full picture of what the component does. I think this is important.
I'll keep testing some more and let you know if i find something.
Thanks,
-A.

**Comment by ayezioro** [comment]
Forgot to mention: if you add the explanations to the hint, you can also add some explanation about the Solar Volume, which is the resulting space between both envelopes. I will also give a reference to the original paper Guedi published (at least).
-A.

**Comment by boris-p** [comment]
Thanks @ayezioro, i'll try to address the points one by one - 

Running time - it is a process that takes time, we've made many efforts in improving performance, both optimizing the code and allowing for parallel processes to run, of course there are many things that could be done to improve upon this, but I think that at least for now there's a good trade off between performance and usability. The times really depends a lot on both the input parameters like the list of suns and the surrounding geometry and on the grid size. I agree that it's crucial for users to know how to set the initial parameters and will prepare a tutorial to walk people through the process, another thing that effects the running time a lot are the parallel_ and _numOfCPUs_ parameters and setting parallel_ to true and trying different values for _numOfCPUs_ could improve performance a lot.

Changing "_obstacleCrvs" to "_context"- The idea behind making the user pick curves and not automating the process was done for a few reasons - 
-To improve performance (avoiding as much as possible irrelevant or unclean geometry). 
-Making the user more conscious of the choices and their effect on the simulation.
-Allowing for greater control for the users - for instance if we want parts of buildings to be not shaded just from the first or second floor to get a bigger solar envelope, it's easier to adjust using individual curves.
*  I'm not sure I completely understand what you mean by planar curves, curves on the xy plane? curves are exploded into segments but so far the component was only tested with curves without a changing z value, although it's a good idea to think about testing and supporting more elaborate curves.

The comparison between the older and newer versions - the reason that the points are not in the same order is the way the parallel processes are running, generating the point list in a different order.    

change "SolarRights" to "SolarEnvelope" to be consistent to the input - consistent to which input? the idea is that they are both solar envelopes, one represents the rights and the second the collection. I'm using the terms defined by Guedi and E.Shaviv here - http://www.ibpsa.org/proceedings/BS1999/BS99_C-22.pdf

Sure, adding some more information for the users would be great, sorry for the ignorance, but what do you mean by the hint of the component?

**Comment by ayezioro** [comment]
I've been checking the SE results and i have some thoughts, specially regarding the Solar Collection option.
See attached: https://www.dropbox.com/s/cxj9jc6bmyubmuo/solarEnvelopeAdvanced_ToReport.gh?dl=0
Please use it as it is defined (only one hour), for understanding the issue i'm writing here. The SolarCollection result shows a planar mesh and in my opinion it has to be in a much lower leve and not planar. I say that based on the view from the sun component that i added to the example. The SolarRights look fine (with nice changes becouse of the streets). But the Collection is not expressing any change. For this example i used many curves for the Collection, and this may cause some difficulties ...? The SolarRights also have a lot of curves but they are managed well (so far).
The attached file will run right away after opening. But don't worry. It takes just a few seconds :-).
Your response get m ein the middle of the writing. So here are my answers to yours:
-. The hint is the message that appears when the mouse is hovering over the component icon.
-. Can accept your reasons for inputting curves, rather than breps.
-. Solar Rights vs Solar Envelope. You are right. The ValueList component in the example should give "Solar Rights" and "Solar Collection". But in the input envelopeToRun_ the hint should be consistent with that also. It isn't.

Thanks,
-A.

**Comment by boris-p** [comment]
@ayezioro , i'll elaborate on the hint part so it would be more explanatory for the users and have changed the hint next to the envelopeToRun parameter. 

Thanks a lot for catching the problem with the solar collection, had a few small bugs there, could you take a look at this file ? I think it should work now
https://www.dropbox.com/s/fak48pk22gju27h/solarEnvelopeAdvanced_ToReport_boris_5.8_4.gh?dl=0

Taking a closer look at the file though, I saw something strange with the created polysurface  - 
![1](https://cloud.githubusercontent.com/assets/3728381/9089134/024d23e0-3b94-11e5-88c3-d88ab720a3f2.png)
![2](https://cloud.githubusercontent.com/assets/3728381/9089161/0a5dff64-3b94-11e5-9524-f6fae635b54a.png)
![3](https://cloud.githubusercontent.com/assets/3728381/9089167/10212e76-3b94-11e5-8696-a519f1fb98cc.png)

seems like it breaks when the z differences are too big. A closer look at the surfaces shows a strange situation - 
![4](https://cloud.githubusercontent.com/assets/3728381/9089231/63cd3768-3b94-11e5-89e1-696c2a084213.png)

I'll take a look at it a bit later, @chriswmackey   maybe we should open it as a separate bug?
Or just continue the discussion regarding all of the issues of the component on this thread?

**Comment by ayezioro** [comment]
Hi Boris,
Getting better. Try now with 5 hours instead of 1. You'll see very "nice" holes in the SRE. Maybe is a Tolerance issue?
As for the SCE i'm still trying to digest that what you get is right. Let me think ...
I think the discussion can remain here for now. As Chris suggested above, until the release. I don't believe many more issues will arise anyways.
One more thing. For some reason Rhino crashed badly. I wasn't in the middle of something but was testing your last file. Can be related, maybe yes, maybe not ...
-A.

**Comment by boris-p** [comment]
@chriswmackey , could you help me out with something? 
Looks like there's some problem with the rebuilding of the new polysurface in the computeFinalSolarVol method. I'm wondering if instead of this part - 

#Convert the faces of the new mesh over to a brep.
        topBrepList = []
        for faceCount, face in enumerate(self.regionMesh.Faces):
            if face.IsQuad:
                faceBrep = Rhino.Geometry.Brep.CreateFromCornerPoints(Rhino.Geometry.Point3d(self.regionMesh.Vertices[face.A]), Rhino.Geometry.Point3d(self.regionMesh.Vertices[face.B]), 
                Rhino.Geometry.Point3d(self.regionMesh.Vertices[face.C]), Rhino.Geometry.Point3d(self.regionMesh.Vertices[face.D]), scriptc.doc.ModelAbsoluteTolerance)
            if face.IsTriangle:
                faceBrep = Rhino.Geometry.Brep.CreateFromCornerPoints(Rhino.Geometry.Point3d(self.regionMesh.Vertices[face.A]), Rhino.Geometry.Point3d(self.regionMesh.Vertices[face.B]),
                Rhino.Geometry.Point3d(self.regionMesh.Vertices[face.C]), scriptc.doc.ModelAbsoluteTolerance)
            topBrepList.append(faceBrep)

 we could just use this - 
Rhino.Geometry.Brep.CreateFromMesh(self.regionMesh,True)
 
I've tested it with the example file @ayezioro has provided and it seemed to produce holelesser geometry, although should test on a more complex base surface.
  File attached - https://www.dropbox.com/s/onbd2e4goego13t/solarEnvelopeAdvanced_ToReport_boris_5.8_5.gh?dl=0

Allthough an envelope like this is pretty extreme and is created by having a very limited amount of suns, it's good to catch these things.

@ayezioro, do you remember exactly what you were doing that made the rhino file crash? Or could you reproduce it?

**Comment by ayezioro** [comment]
Ok. I've been thinking ... and testing. The best way to see this is one hour at a time and look from the sun. Then you get it right. So now i can say the component is working great. Just the voids in the surfaces.
So far i tested the SRE with curves on the same level. I'll test different height. But i know that this is only for my fan.
Question: I'm getting the lowest height of -20, without setting the value for the SCE. Is this an internal limit? The hint says that it should be -100 below average height.
For now, that's it.
-A.

**Comment by ayezioro** [comment]
No. Can't reproduce. Maybe the computer was tired :-)

**Comment by chriswmackey** [comment]
@boris-p  and @ayezioro ,

I am glad that the discussion has been really picking up and I am sorry for being away for the last 3 days.

I just wanted to say first that I have merged the old pull request, @boris-p , and I have added the example GH file from this pulled version to the Ladybug Example files in our app.box.  Anyone can now download it from here:
https://app.box.com/s/aormi5g1n9cs47brhmrz1hqnf5rr0lul

@boris-p , I am not getting any holes in the brep for the most recent GH file that you posted and using the Rhino.Geometry.Brep.CreateFromMesh method is certainly much better than what I had written before.  I am not sure if you still need my help but let me know if I am not understanding.  Feel free to make a new pull request once you feel we have gotten everything ironed out here.

I agree with all of @ayezioro 's comments and I will just add one last request for the simplicity of the component: now that we have the numberOfCPUs input, I don't think that we need a parallel option since users can not run it in parallel just by setting the number of CPUs to 1.  Let me know if you agree.

I'm looking forward to hopefully moving this out of WIP soon!
-Chris

**Comment by ayezioro** [comment]
Hi Boris and Chris,
Related and not related.
Check this one: https://www.dropbox.com/s/6i6qrhgtlem24z6/solarEnvelopeAdvanced_ToReport_boris_5.8_5_AY.gh?dl=0
I'm calculating the radiation of the envelope in LB. I don't understand why there are patches where the radiation is 0. At first i was thinking the normal was facing the wrong direction. But showing them i see that all of them are fine.
Any ideas?
-A.

**Comment by chriswmackey** [comment]
@ayezioro ,

I do not yet know the best way to fix the code but I do know why it is happening and I have two temporary workarounds.

The problem is happening because some of the brep faces created by the solarEnvelope are doubly-curved in a way that seems to cause the analysis points to be generated slightly below the analysisSrfs.  The two workarounds are either:

1) Triangulate all of the faces of the brep before connecting it to radiation analysis, which ensures that all generated faces are planar (thus avoiding double-curvature):
![triangulated mesh](https://cloud.githubusercontent.com/assets/5567574/9149131/95a1be3c-3d68-11e5-9529-ddc6ef9f406a.PNG)

2) Increase the distFromBaseSrf to a value that ensures that all generated points are far enough away from the base srf that they are not accidentally located underneath it:
![increase distfrombasesrf](https://cloud.githubusercontent.com/assets/5567574/9149128/90bad570-3d68-11e5-8ed3-6e9e5f7b172c.PNG)


Since the radiation component is @mostaphaRoudsari 's, I feel that the decision is up to him whether he wants to address this type of case in the code and how it should be done.  @mostaphaRoudsari , I am happy to help implement one of these options if you think it would be worthwhile.

Here is a GH file showing how to triangulate the brep for the first workaround:
https://www.dropbox.com/s/hl8bt9abzo0a9pa/solarEnvelopeAdvanced_ToReport_boris_5.8_5_AY_CWM.gh?dl=0

-Chris

**Comment by ayezioro** [comment]
Hi Chris,
Seems to be that the triangulate option works better than changing the distance from base. At least from the test i did. So at least we know how to deal with the problem now.
Thanks,
-A.

**Comment by boris-p** [comment]
hey @ayezioro and @chriswmackey . Following the discussion in the last couple of days, I've incorporated the bug fixes and improvements that you have suggested. @chriswmackey , I agree regarding removing the parallel input.
The only open issue for me at this point are the description texts, I think that for people with familiarity with solar envelopes this would be easy to understand but tricky for people who don't have background knowledge and also tricky to explain very concisely so while thinking of a better and more descriptive text, the priority for me now is to make a more detailed tutorial (hopefully would get it done in the next few days). 

Before updating and making a new pull request, I wanted to ask if at the moment there are any more improvements / suggestions / bug fixes from your side?

**Comment by ayezioro** [comment]
Hi Boris,
I have no new comments.
As for the description i don't think a complete/comprehensive text should be written. Some responsability must be in the hands of the users. You can never guarantee the good use of any tool. But i think that a reference to a paper can be more helpful. For instance you can mention some Ralph Knowles (he is the one who proposed the term) and of course Guedi's.
-A.

**Comment by mostaphaRoudsari** [comment]
Hi @ayezioro! thanks for finding this new bug. I would open a new issue for the bug with radiation analysis mesh issue. If you agree I won't put a high priority on this.

@chriswmackey there is no such a thing as someone's component. Feel free to go ahead and get it fixed.

PS: Nice discussion here btw.

**Comment by ayezioro** [comment]
Hi Mostapha!
Of course: give it your own priority. 
-A.

**Comment by boris-p** [comment]
hey guys, I've just sent a pull request containing the latest version with the bug fixes for all the known issues, also uploaded an updated an a bit more comprehensive example file - https://www.dropbox.com/s/ptex3cs22vtp4xw/solarEnvelopeAdvanced_13.8.gh?dl=0

Regarding the description or a reference to an article, I've decided to leave it as it is for now. 
The reasons beings that there is a description at the top of the component explaining the basic principles and as I'm not utilizing any specific solution but rather a combination of a few approaches with a few new additions, I think that referencing a specific article might be a bit partial or misleading (btw there is a reference to sustarc and Guedi as it was the base for the component logic). 
For users who are interested in the principles behind the component, a simple google search of solar envelope leads to a lot of information, and of course i'm working on a detailed tutorial in which i'll explain both the component functionality and the theory behind it.

If it's possible, i'd be great if you could take a quick look at the component to see that in my changes I have not messed anything up, and as I've said a tutorial is also on the way.

/b

**Comment by ayezioro** [comment]
Boris,
The example file shows that the component is dated July 6, which is the previous version i have. Is that right? In the github there is no yet an updated version.
So i'll wait for your confirmation before checking.
-A.

**Comment by ayezioro** [comment]
Disregard the previous. I see from the content that they are different.
Runtime is almost the same as the previous. So it is OK.

Didn't find any other issues. It is fine now with me.
Though, still think that references should be added. If you combined different approaches then a reference of those is in order. People using this component in the future can not guess which approaches are those, and they will disregard completely other's work, and worse, will think that this is a completely original work. A video explaining the component doesn't cover this vital information.
-A.

**Comment by mostaphaRoudsari** [comment]
@boris-p Thanks. It looks ready to go. I second @ayezioro's comments about referencing. I think it's the right approach to reference their work explicitly. What do you think?

**Comment by boris-p** [comment]
I'm convinced, thanks @mostaphaRoudsari  and @ayezioro   :)
Updated the pull request (I think, the component should have today's date) and also the example file - https://www.dropbox.com/s/qda0vhpga1zotzs/solarEnvelopeAdvanced_14.8_as_example_file.gh?dl=0

Will update in the next few days regarding the tutorial

**Comment by mostaphaRoudsari** [comment]
@boris-p Great! Thank you. :) I will check and merge it after work this evening. Cheers. :beers:

**Comment by ayezioro** [comment]
Thanks Boris. Looks good ... and fair.
-A.

**Comment by chriswmackey** [comment]
@boris-p ,

Sorry for being AWOL this past week.  I was just starting a new job and couldn't check everything until the weekend.

I agree with all changes that have been made and it is really looking great now.    I feel safe saying that, when we come to the next stable release very soon, we should be able to take it out of WIP.

I merged your new pull request and updated the example file on the app.box link here:
https://app.box.com/s/aormi5g1n9cs47brhmrz1hqnf5rr0lul

Looking forward to the tutorials!
-Chris

**Comment by ayezioro** [comment]
Hi,
I'm trying the component right now and suddenly come to my mind a question: Is the _obstacleCurves a MUST, tat without assigning them the component can't run? Isn't the base curve enough for that?
The reason i'm asking is that, in fact, there can be cases where i don't have any neighboring buildings that can affect the envelope.
Makes any sense? I think so, but what do you think guys?
-A.

**Comment by ayezioro** [comment]
Hi again,
I'm still thinking and i'm coming to the conclusion it doesn't makes sense. If there are no neighbors, or any boundary where to refer, so anything goes and you can build as high as you can.
I'm not deleting the comment, just as a reminder if me or someone else will raise the question.
-A.

**Comment by ayezioro** [comment]
Hi,
Sorry for so many interventions here ...
I've been checking last Chris's change comparing with the previous one from Boris (yesterdays). I'm coming to the conclusion that the "best" thing is to leave Boris's. Please see the images attached. The first one before the change and the second, after the change.
In the first one you can see (from the sun point of view) nicely how the envelope fits the voids between the buildings and their limits. In the second something is happening that causes a "buggy" envelope where parts of it are not fully complying with such conditions.
![beforechrischange](https://cloud.githubusercontent.com/assets/7526023/9292819/cf755ebc-4415-11e5-9e9e-c19720a6a819.jpg)
![afterchrischange](https://cloud.githubusercontent.com/assets/7526023/9292820
/d37abc64-4415-11e5-936b-ee469fd06ac6.jpg)

Sadly, i think that the best option is just to warn the user to give a bigger distance from surface.
-A.

**Comment by ayezioro** [comment]
And the very last (for today, hopefuly).
See attached: https://www.dropbox.com/s/2b4wf5tjviidg0x/SolarFan_solarEnvelope_Chris%2BSaeran%2BBoris.gh?dl=0
I'm comparing the new envelopes with the previous/other versions.
So first thing i noted is that the SolarEnvelope component is not there anymore. I'm using some version i have. If you see the outcome of the Advanced with this one the results are pretty much close.
Comparing the Advanced with Saeran's version there are pretty much discrepancies. Some parts of the envelope are shifted. I'm quite sure i'm simulating what is needed from the amount of hours and months concern, so both envelopes are simulated for the same periods of time. Maybe @saeranv can comment on this?
I think is nice to make such comparisons, just to be sure about results, etc.

Solar Fan and Solar Catch are not comparable unfortunately.
-A.

**Comment by saeranv** [comment]
Hey everyone, I need to catch up on this thread but I'll look into the discrepancy and figure out what's happening. This new component look pretty sick.

S

**Comment by chriswmackey** [comment]
@ayezioro ,

Thanks for all of your tests.  To clarify, are you suggesting that we should not triangulate surfaces in the radiation analysis component but just give a warning if a point is generated below a surface?  If so, please post the suggestion quickly on the issue I opened:
https://github.com/mostaphaRoudsari/ladybug/issues/167

The comparison of the envelopes is interesting.  Overall, they are pretty similar but my instinct tells me that the reason why Searan's might be an outlier has to do with how he is defining the solar geometry.  In Searan's component, i see that he is using the sun position points instead of sun vectors.  Perhaps this might cause some distortion in the geometry if the sun path scale is not large enough to make the points far away?
![getting sun points](https://cloud.githubusercontent.com/assets/5567574/9295136/757bcf42-4431-11e5-838b-11e372d3fb61.PNG)

Looking forward to your insights @saeranv .

-Chris

**Comment by ayezioro** [comment]
Hi Chris,
I posted in the #167.
I think i understand now the differences with Saeran's. 
His input _requiredHours is not clear. In the example i sent i gave 5, hoping the hours will start from 10 to 14 (5 hours, including 10 and 14). Checking his code i see that the hours are 9 to 14. Which are 6 hours instead. Now the results are much closer. Just some ridges that i can't understand. But out of them mostly it makes sense. 
So it is nice that sun geometry is consistent, even with different approaches :-)
For Saeran's component i'll recommend to explain better the number of hours input to the user.
-A.

**Comment by saeranv** [comment]
Hi guys, 

I spent some time today looking through the solar envelope examples, and there's a couple of issues that are causing the discrepancy. 

1. @chriswmackey  I checked the solar points versus solar vectors; and it seems to yield the same results. So I don't think this is the problem here. It's helpful to know that you're using a different method though.

2. As @ayezioro  mentioned, one of the issues is the interpretation of _requiredHours. Essentially @ayezioro interprets 5 hours as, [10,11,12,13,14], but I interpret this as 4 hours (there are four 60min intervals between 10 to 14). Therefore my code interprets 9-14 as 5 hours. It's a funny semantic ambiguity that I agree should be clarified :) .

3. Thirdly, the time periods at which I'm generating the solar vectors/points is not the same as in Chris' component. In my component, once the user submits the 5 hour period, the component calculates solar noon (the time at which the sun is highest in the sky) and calculates the hour range where solar noon is the midpoint; i.e. for five hours it starts at 2.5 hours before solar noon and 2.5 hours after solar noon. Every month has a different solar noon (somewhere between noon and 1pm) so every month solar extrusion has a different time range.

This method ensures the derived solar envelope is extruded at the precise five hour period for each month where the solar geometry will give us the greatest volume. The idea is that for a non-specialist user who doesn't understand the specific variations of the solar positions, it's better to build this intelligence into the component. 

In the example file here: https://dl.dropboxusercontent.com/u/4598766/SolarFan_solarEnvelope_Chris%2BSaeran%2BBoris_SV_edit.gh, you can see that I added a modified version of my component without this solar noon calculation, and you can see it is exactly the same as @chriswmackey's solar envelope. As for my original envelope you can see that once we make the hour ranges consistent, my component has a greater volume than Chris', because it's using the solar noon calculations. 

4. Finally, there was some minor variations caused by the minimal selection of solar vectors in my component, which was an attempt to reduce computational processing time. I looked at it again, and decided it wasn't worth the inaccuracy and rewrote it to include more solar vectors to reference the solar envelope. So the component is slightly slower, but more accurate now. 


So I think that's all the issues. I didn't realise that there were several different assumptions between our two solar envelope and fan components. I think this might be a good opportunity to differentiate both, with mine serving as a 'simple' or beginner's version of the two components, because it builds in a lot of data relating to the various key solar positions.

As always, thanks for pointing out the discrepancies @ayezioro; it was very helpful. 

S

**Comment by ayezioro** [comment]
Good!
@saeranv  i would add some brief explanatior or example in the hint of the requiredHours, so users can fully understand the meaning of the input.
The only last think to fix is that the Advanced Envelope is @boris and not @chriswmackey. So the credit goes to Boris (and probably a bit also to Chris if i'm right).
-A.

**Comment by chriswmackey** [comment]
@saeranv ,
Awesome!  Thanks for the rigorous testing! 

@boris-p 's solar envelope can model the exact same case as @saeranv 's component by using the new solarTime feature that I added to the sunpath a month ago, which gives you solar noon and any variation off of that.  The two are matching exactly now:
![solartime](https://cloud.githubusercontent.com/assets/5567574/9318166/23e86d64-450f-11e5-84d5-2f8e6f89bad1.JPG)

And, yes, thank you @ayezioro for clarifying.  @boris-p 's awesome robust solar envelope component is replacing the old (incredibly buggy) one that I helped make and most of that old buggy one was derived from @mostaphaRoudsari 's shadingDesigner code anyway.  So I am the lowest of those who deserve credit here :)

-Chris

**Comment by ayezioro** [comment]
Hi Chris,
I'm seeing the definition (hint) in the solarOrStandardTime. Are you sure that True is solar time? I think is the opposite. The way botth envelopes coincide is setting the input to False (or not setting it at all since it is the default). I understood from @saeranv that he is using solar time. Moreover, the sunpath of the True option looks as Standard Time (i.e. No analemas).

![capture](https://cloud.githubusercontent.com/assets/7526023/9322650/1b4a58e6-457f-11e5-8b11-60a72fea291f.jpg)

-A.

**Comment by boris-p** [comment]
@chriswmackey , i'd give you a bit more credit on the help with the component, a lot more actually.

@ayezioro in relation to one of your earlier comments, regarding what happens when we don't assign obstacle curves with solar rights. When Chris and I have spoken about it in the past we came up with automatically calculating a self envelope (taking the _baseSrf as the border) and that actually stopped working when I updated the parallel options, luckily the fix was easy so in the new version this would work.
What do you all think about solar Collection, maybe we should add the same option (taking the _baseSrf when no obstacle curves are available?) I'm not sure what would be more intuitive and less confusing for the user, might be better to leave it as it is.

Regarding the long awaited tutorial, I've recorded a video  - https://www.youtube.com/watch?v=hRtsxgJCVEw 
It would be really great if you could take a look and let me know what you think. Weather it is clear, if I skipped anything vital, if my voice is annoying. Really, I not easily offended and any comments and criticism would be greatly appreciated.

I'll update the description and add links to everything, anything you would recommend me adding or maybe there's a standard format for the existing ladybug tutorials? 

Regarding the tutorial file, i'll update it to support the latest code changes. In addition, do you think that it might be easier for the users if we have all the geometry saved in a 3dm file instead of internalizing it? As with some of my previous comments, if there's already a standard way you're used to doing these things with other components i'll be glad to adjust the example file to fit the standard. 

That's all from my side, thanks a lot everybody!

**Comment by ayezioro** [comment]
Hi Boris,
For the obstacle curves i understand it is not working in the version we have (or i have). But it is good you it can work without it. 
The name of the baseSrf should be changed to baseCrv (right?).

As for the video. This is great, but ... since you don't mind the critics. Here i go:
It is tooooo loooooong. Put yoursel a limit of up to 7 minutes. 
The first 2 minutes you can skip. You don't need to explain about LB and such. Since the video will be in the context of others, the basics are pretty much covered already.
I like the cases when you change the curves location. This is a good demonstration of the envelopes potential. I like much less the obvious things like grid size, CPU, sun vectors and the outputs.
Even though you explained at some point that you use curves and not breps, you need to stress it a bit more. The way i thing is that you'll hide the breps and leave only the curves. After inputting them you can show the breps again.
This is what i have for the video. Now i have some requests, maybe for future versions: a bake option where you can send the envelope to Rhino to a new layer named according to the analysis period. Of course it can be done now with the output, but i can ask, right? I'll ask for this also to @saeranv.
Sorry if i've been to picky and thanks,
-A.

**Comment by mostaphaRoudsari** [comment]
Hi @boris-p! Nice work. I actually liked the video. I agree with @ayezioro's comment that it might be a little bit long but people can always move it forward if they think they know some stuff. There will be people who will be watching the video and won't know about Ladybug. I would even suggest to add a link to the forum in video's description.

Changing baseSrf to baseCrv makes sense to me. It can be very confusing for users.

I think we should internalize the geometries in the example file, and also make sure visualization of the file looks good once you open it. That's how we have been doing it and it saves us so much time.

I also have a suggestion. My suggestion is to automate the process of selecting curves in your file. I'm pretty sure some users will spend a good amount of time re-drawing the edges of the surfaces. Here is an example on how it can be automated:

![getcurves](https://cloud.githubusercontent.com/assets/2915573/9346721/4fed7690-45ee-11e5-9466-757d820d7316.png)

**Comment by chriswmackey** [comment]
@boris-p @ayezioro and @mostaphaRoudsari ,

To clarify the baseSrf input (as I understand it), it uses the Brep type hint such that it is automatically converting any planar curve that is connected to a planar Brep before running it through the component.  Personally, I think this this feature is great because it allows you to run planar curves or planar surfaces easily through the component but you can also plug in an complex surfaces representing uneven terrain (something that I find myself doing fairly often, actually).  Perhaps this awesome range of possible inputs could be clarified in the video.

The only major feature that I see as important before moving the component out of WIP is to have the option of making the self-bounded solar envelope if no obstacle curves are connected.  I really liked this feature in the older version and I think that it is not fair to replace the old solar envelope component without providing support for this feature.


I agree with all comments about the video and I also agree with @mostaphaRoudsari that it does a good enough job right now so editing it could probably be pushed to a lower priority over the self-bounded envelope.

Geometries in the example files should definitely be internalized.  It just makes sharing the files so much easier.

Automatically pulling off the bottom and top curves of any closed breps (as @mostaphaRoudsari suggests) sounds like a very useful part of the workflow.  I imagine that this might be best done with another component that we put into the "Extra" tab since trying to automate it inside the solar envelope component might restrict the freedom to input any curves that you think are useful (and I can also imagine that the automating function could produce no top or bottom curves for a given brep if the "max Angle" is not great enough).  Let us know if you were imagining otherwise @mostaphaRoudsari .  If not, I it seems that this can also be something that is not critical to bringing the component out of WIP.

Finally, I agree that a bakeIt option would be useful, although I would also say that this feature is not vital to bringing the component out of WIP.  After all, I still have to implement this on a ton of components that I have made (https://github.com/mostaphaRoudsari/ladybug/issues/104).

-Chris

**Comment by boris-p** [comment]
Regarding baseSrf input, @chriswmackey  is entirely correct, the component works with closed curves, surfaces and more complex breps, I should have done a better job of explaining that in the video and already updated the example file so that the input matches the name.

@chriswmackey and @ayezioro , just to be clear,  the self-bounded solar envelope if no obstacle curves are connected is already functional, it didn't work because of a recent bug due to one of my updates but is now fixed, it's just a matter of sending a new pull request now.

@mostaphaRoudsari and @chriswmackey , following your comments I internalized all of the geometry, also tried to simplify the GH definition as much as possible and used @mostaphaRoudsari's suggestion and the honeybee component to automatically extract the top and bottom borders from the buildings. I think that it is good to keep the input parameters for the component as curves and not to automate the process more than that, I have elaborated on the reasons on this thread, a copy of what I wrote - 
"The idea behind making the user pick curves and not automating the process was done for a few reasons - 
-To improve performance (avoiding as much as possible irrelevant or unclean geometry). 
-Making the user more conscious of the choices and their effect on the simulation.
-Allowing for greater control for the users - for instance if we want parts of buildings to be not shaded just from the first or second floor to get a bigger solar envelope, it's easier to adjust using individual curves."

The latest example file - https://www.dropbox.com/s/x7ym2apso6xmvl0/solarEnvelopeAdvanced_23.8_example_file.gh?dl=0
Anything else you would change there? It's automatically set to run, maybe it's better to set it to false and let the user run it? although it should take just a few seconds to run.

Regarding baking, I could add this functionality into the component, but I also see benefits in having one baking component which could be used by all elements (maybe taking geometry and layer name or layer hierarchy as inputs). I'm afraid I might be not updated and maybe there's already a discussion or a smarter solution than what I would implement solely for solarEnvelope. If this is a WIP and relevant i'd love to pitch and contribute to the discussion and/or the development. 

Lastly I have a suggestion. From my side, I think that the component is stable enough to be released and that example file and video, although could be improved, explain the possible component options and usage. I would release it, this would allow for more people to try it out, maybe find some bugs or give suggestions for improvements, and following the new input I could update the component and then also the video, taking note of all of your (very appreciated) comments.
If you're okay with that, i'll send a new pull request and update the video description and add links to all relevant resources. Besides that, when you'll give me the green light, we could think of a message to write on facebook, or maybe it's better to wait for a new release?  

-boris

**Comment by ayezioro** [comment]
Hi Boris,
The component in the example looks and works fine. I tried inputting surfaces and curves 9at the same time) and it work OK. Nice :-)
Question: The baseSrfCrv need to be planar? Or they can be 3D? I assume that planar, so a short explanation about this in the input hint will be nice.
For the specific example i like less that the geometry of the plot crosses the topography. So there is a part below ground that you are asking to keep the solar rights. BTW, this is the reason why i asked the previous question. I would change the geometry of the plot or change the topography. But this is me ...
As for the input names now that they accept crvs and breps y suggest to change them to baseSrf_Crv and obstacleSrf_Crv (or something similar. Maybe baseGeo and obstacleGeo ...).
Agree with your comments. regarding the use of HB component to filter ground/roof. It is very important to give the freedom to change the curves location (or at least some of them).
In my opinion the component is robust enough to set it free from the WIP to the mayor ligue :-)
The bake option, though i suggested it, is not priority #1, and since there is an intention to make something more comprehensive indeed no need to deal with this right now.
Thanks,
-A.

**Comment by mostaphaRoudsari** [comment]
Hi all,
I agree with @boris-p's comments. The component is ready to go. I think we don't really need to have a component for extracting the curves or automate it inside the component. Let's leave it to the users and trust that they know better what they need as the input.

@ayezioro curve to brep functionality is an automated Grasshopper functionality. Curve needs to be planer in this scenario.

![image](https://cloud.githubusercontent.com/assets/2915573/9429175/9d834e8e-4995-11e5-9c64-0354c9f68c5c.png)

@boris-p go ahead and send a pull request. Let's get more people involved. Thanks.

**Comment by ayezioro** [comment]
Hi Mostapha. The reason i asked is not related to GH functionality but the SE's. Since i see in the example that the input is given by a surface and not by a curve (as previous example files) i was wondering if non planar curves or surfaces are "legal" for the component. I understand that no. So, fine.

Just tested that the SR option works without obstacles, thus given the functionality of the previous SolarEnvelope. Very good.
-A.

**Comment by boris-p** [comment]
@ayezioro  baseSrf_ accepts curves, surfaces and breps.  It doesn't have to be a curve so I think that changing the name to a crv or restricting the parameter, forcing it to be a curve might be limiting to some users or some workflows.
obstacleCrvs on the other hand has accept curves, it's not just a matter of convenience because sometimes we might want to define single lines and not closed polylines. 
So the idea is that while we are pretty general in baseSrf_ and accept many options we want to be very specific regarding the obstacleCrvs. I hope that I understood your comment correctly and was able to clarify, 

@mostaphaRoudsari  and @chriswmackey , i'll start writing a short description to post on the facebook page and a longer one with more explanations, images and references to post on the ladybug forum . Once that would be ready i'll upload it here to get your thumbs up, anything else i'm forgetting?

**Comment by mostaphaRoudsari** [comment]
@boris-p all sounds good! I don't think there anything else. I merged the latest version. It should be all good to go now.

**Comment by ayezioro** [comment]
Hi,
Downloaded.
Issues (sorry, but they are very small): it is still in the WIP and is dated 18.8.
Don't hate me.
-A.

**Comment by mostaphaRoudsari** [comment]
Hahaha! You're the best Abraham. Is the issue in github or the example file?

**Comment by ayezioro** [comment]
In the github. Inserting as new.
Thanks! :-)
-A.

**Comment by mostaphaRoudsari** [comment]
That's the right version though. I didn't change the date as this is what
Boris had in pull request

**Comment by ayezioro** [comment]
:+1:

**Comment by boris-p** [comment]
Ooops...
@mostaphaRoudsari  should I update the component and send another pull request, what would be the easiest option for you? (sorry for all the pull requests, I always seem to forget something...) .

I guess it also makes sense to update the code in the example file, even though it's not affecting the functionality.

**Comment by mostaphaRoudsari** [comment]
I would say it is all fine for now. Let's go ahead with what we have and
next time you have more changes you can apply these two changes and send a
pull request. None of them is critical at this point. :)

**Comment by boris-p** [comment]
Hey, just finished writing messages for FB and Grasshopper3D and wanted to run them by you before posting.
If it's all in order, is there a place where you usually upload the example file to, if so could you send me the link and i'll update the message ? 

If there's anything or anyone I left out, forgot to give credit where credit is due or anything else, please let me know so I could fix the wrongs.

Other then that from my side we're good to go!
------------------------------------------------FOR FACEBOOK -------------------------------------------------
Hello Ladybuggers!
We have just released a new solar Envelope component for Laybug called SolarEnvelopeAdvanced! The component offers two envelopes, solar rights and solar collection, which perform a simulation of the maximum area in which we can build without violating the sun rights for surrounding buildings and a simulation of the minimal heights from which we can receive solar access, respectively 

A video tutorial is available here -   https://www.youtube.com/watch?v=hRtsxgJCVEw

An example file  - _____insert example file link____

And a more detailed description and discussion - 
_____insert grasshopper3d link here____

------------------------------------------------------------------------------------------------------------------------
-------------------------------------------------FOR GRASSHOPPER3d ---------------------------------------
Hello Ladybuggers!
We have just released a new solar Envelope component for Laybug called SolarEnvelopeAdvanced! The component offers two envelopes -

Solar rights - A 3D envelope which defines the maximum height of built volume which would not violate the solar access for defined surroundings given filtered sun vectors.

Solar Collection - A complementary 3D envelope which defines the lowest heights from which a new built volume would receive solar access, given a built context and filtered sun vectors.

A video tutorial is available here -   https://www.youtube.com/watch?v=hRtsxgJCVEw

An example file  - _____insert example file link____

The component takes as an input a filtered list of sun vectors, a base surface which represents the site border and a list of curves, representing the surrounding context 
  
![top](https://cloud.githubusercontent.com/assets/3728381/9563440/0a44ec92-4e84-11e5-9e8d-ecb373393c9a.jpg)

and outputs a point cloud and a polysurface -  

![persp](https://cloud.githubusercontent.com/assets/3728381/9563442/1e2b350e-4e84-11e5-91ea-de521f19d5c9.jpg)

which in turn can be closed and further evaluated (On the left - solar collection, on the right - solar rights) - 
![4](https://cloud.githubusercontent.com/assets/3728381/9563445/2b6fc90a-4e84-11e5-99bc-79e5ee07dea3.jpg)

By Boris Plotnikov and with the assistance and guidance of Prof. Guedi Capeluto, based on SustArc model. The component could not have be alive without the invaluable help of the awesome Chris Mackey and Mostapha Sadeghipour Roudsari in development and integration, also a big thanks is due to Abraham Yezioro for testing and advising on usability.

For further reading it might be worth taking a look at Ralph Knowles's work, e.g - http://www.fau.usp.br/aut5823/Acesso_ao_Sol/Knowles_2003_Solar_Envelope.pdf
 and G. Capeluto and E. Shaviv's, e.g - http://www.ibpsa.org/proceedings/BS1999/BS99_C-22.pdf
the component relies to a great extend on the concepts described there

Thanks for reading and enjoy!

**Comment by mostaphaRoudsari** [comment]
Hi @boris-p!

It looks very good. Explanations is clear and nice graphics. My only comment would be to add a line or two and mention the enhancements and new potentials that the new component brings to table. In other words, you want to be clear about **why** people should test this one today and stop using the old ones?

Back to download links we're still not sure what would be the best structure. Until we figure that out a dropbox link should be fine.

BTW you know it's never final until Abraham says so! :) @ayezioro are we good to go?

Thanks again @boris-p! Looking forward to see you close this issue. :)

**Comment by ayezioro** [comment]
Hi Mostapha,
I'll take your comment as a compliment :-). Big responsibility!!
One question before the final comment: Why the name of the component needs to be SE Advanced? I'm asking this because the previous "old" one is not distributed any longer. It is not that there will be 2 of them (besides of course Saeran's).
I believe that beside this question i give a BIG blessing for this great component. I know the community will find it more than useful. I'm glad that the credits are given.
I will send a pull request that takes the component out of the WIP.
-A.

**Comment by mostaphaRoudsari** [comment]
Hi Abraham! Thanks. With no doubt it is a compliment. It's actually a big one. We are very lucky to have you on board.

I think you are right and we won't have another solar envelope component. At some point @chriswmackey removed it from the package. I would say it makes sense to remove advanced from the name. At the same time it might make so much extra work for @boris-p. I would say we can fix this issue before the next official release so we don't stop Boris from the current post.

**Comment by ayezioro** [comment]
Made my day :-)
100% with your proposal.
-A.

**Comment by boris-p** [comment]
Great ! @mostaphaRoudsari and @ayezioro , regarding the advanced, I agree that it makes little sense now that it's replacing the old component, and would be nice to have a simpler name but for now I would keep it given that the component, the example file and especially the video tutorial are already prepared. I don't want too big of a discrepancy between the tutorial and the actual file (might be very confusing for the users).

@mostaphaRoudsari , I agree, i'll add some more information in the grasshopper3d message regarding the advantages of the tool and will post the messages tonight or tomorrow.

The honor for closing the issue is all @chriswmackey 's, I don't think i've got permissions :)

**Comment by boris-p** [comment]
posted messages on fb and Grasshopper3D - 
http://www.grasshopper3d.com/group/ladybug/forum/topics/new-solar-envelope-component-now-available

fyi

**Comment by mostaphaRoudsari** [comment]
Thanks Boris! Very well accepted by the community. It's very flexible and useful. I will be using it for my class.

I leave this issue open for @chriswmackey to close!

**Comment by boris-p** [comment]
great to hear and see, thanks!

**Comment by chriswmackey** [comment]
Hello All,

I am sorry for my huge absence over the last week and a half but I am so happy with how far everything has come and how everything has turned out.  I barely deserve the honor of closing the issue but I will do it with pride and gratitude to you all.

Congratulations @boris-p !
